N3X15/SpaceEngineers-CE
# Gravity Generator Modding Improvements

# Changelog
## Executive Summary
- Mods based on Gravity Generators can now define a default field size and strength.
- Mods can now re-implement some gravity generator code, if needed.
- Blocks using Gravity Generator behaviors are no longer forced to play the generator's droning sound.
## Details
- Added MyBoundedVector3 for neatly bundling fieldsize bounds.
- Added .GetOrDefault(defaultValue) to floats and SerializableVector3s, making detection of uninitialized MyObjectBuilders a bit more standardized (uses null rather than negative or huge values).
  - Note:  Also detects NaN as a marker to maintain backwards-compatibility for an older version of this PR.
- MyGravityGeneratorBase:
  - Set to public, so modders can inherit it from their own classes.
  - Changed the idle noise to be configurable via `<PrimarySound>`, like reactors.
- MyGravityGenerator (and friends):
  - Set to public, so modders can inherit it from their own classes.
  - `<FieldSize>` (MyBoundedVector3) added to definition and object-builder.
    - Default:
      - Min: <0f,0f,0f>
    - Max: <150f,150f,150f>
    - Default: <150f,150f,150f>
- MyGravityGeneratorSphere (and friends):
  - Made public.
- Added MyGravityProviderDefinition object builders to handle gravity acceleration-related serialization. Can eventually be worked into Gravity Spheres and whatnot instead of the current duplicated code.
  - `<Gravity>` (MyBounds) added to definition and object-builder.
    - Default: Min=-1f, Max=1f, Default=1f
  - Internally, Gravity Spheres do not share the same Gravity scaling as Gravity Generators ([-1,1] for Gravity Generators vs. [-G,G] for Gravity Spheres).  This presents a large problem for backwards-compatibility and will have to be tackled in a later commit.
## Rationale
- Some mods need the ability to adjust the size of the default field.  This change also permits proper power usage scaling and balancing, as maximum and minimum values can be established.
- Modders desiring the ability to change how gravity affects objects can do so without hacking the game itself, which is troublesome for multiplayer support, and a hassle for end-users.
- This changeset also provides backwards-compatibility for older mods, establishing the current hard-coded values as defaults.
- Future changes to block definitions that leave behind out-of-bounds savedata for fields/gravity strength will not cause buggy behavior, as this changeset adds sanity checks to deserialized save data for gravity generators.
# Test Script
1. Create a new mod with a gravity generator block. (I suggest the GravTube mod listed above.)
2. Set the following in the MyObjectBuilder_GravityGeneratorDefinition:

``` xml
<FieldSize>
  <Min x="0" y="0" z="0" />
  <Max x="2.5" y="2.5" z="2.5" />
  <Default x="2.5" y="2.5" z="2.5" />
</FieldSize>
<Gravity Min="-1" Max="1" Default="0.9" />
```
1. Save the mod.
2. Start a new game.
3. Ensure existing gravity generators on the map are at 150m field size in all dimensions and 1G strength, and that those are the maximum values.
4. Open G menu and equip the placer for the modded GravityGenerator block.
5. Place the block on a powered grid.
6. Ensure the placed block is set to 2.5m in all dimensions, with a minimum of 0m and a maximum of 150m.
   - Note:  the values shown in the panel are rounded, CTRL+click on the slider to see true value.
7. Ensure the placed block has 0.9G field strength, with a minimum of -1G and a maximum of 1G.
